File: csharp/robot-name/RobotName.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class Robot
{
    private static HashSet<int> Previous = new HashSet<int>();

    public const int MAX_NAME_COUNT = 26 * 26 * 10 * 10 *10;

    private static Random generator = new Random();

    private readonly char[] ALPHABET = new char[]
    {
        'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 
        'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 
        'U', 'V', 'W', 'X', 'Y', 'Z'       
    };

    public int? ID {get; private set;}

    public string Name
    {
        get
        {
            return TranslateIdToName(ID ??  GenerateId() );
        }
    }

    private int GenerateId()
    {
        int tempID;
        do
        {
            tempID = generator.Next(MAX_NAME_COUNT);
        } while (Previous.Contains(tempID));
        
        ID = tempID;
        Previous.Add(tempID);

        return tempID;
    }

    public void Reset() => ID = null;

    private string TranslateIdToName(int id)
    {
        string reversedName = "";
        for (int i = 0; i < 3; i++)
        {
            reversedName += id % 10;
            id = id / 10;
        }

        reversedName += ALPHABET[id % 26];
        reversedName += ALPHABET[id / 26];

        return new string(reversedName.ToCharArray().Reverse().ToArray());
    }
}

File: python/robot-name/robot_name.py
import string
import random
class Robot:
    robot_names = {0:0}
    MAX_NAME_COUNT = 26 * 26 * 10 * 10 *10

    def __init__(self):
        self.Id = self.generate_id()
        self.name = self.translate_id2name(self.Id)

    def reset(self):
        self.Id = self.generate_id()
        self.name = self.translate_id2name(self.Id)

    def generate_id(self):
        tempID = 0
        while (Robot.robot_names.get(tempID) != None):
            tempID = random.randint(0, Robot.MAX_NAME_COUNT)
         
        self.Id = tempID
        Robot.robot_names[tempID] = tempID

        return tempID

    def translate_id2name(self, id):
        reversedName = [str((id//10**index)%10) for index, value in enumerate(range(3))]

        reversedName.append(string.ascii_uppercase[(id//(10*10*10)%26)])
        reversedName.append(string.ascii_uppercase[id//(10*10*10*26)])

        return ''.join(reversedName[::-1])
